036-manual-step/proposal.md

```diff
@@ -0,0 +1,193 @@
+# Summary
+
+This proposal outlines the addition of a Manual step type to the types of steps
+which can be added to a job.
+
+
+# Motivation
+
+As a pipeline developer attempting to use Concourse to automate infrastructural
+concerns, I wish to preview what my infrastructure tools plan to execute before
+these tools execute said plan. Many infrastructure-as-code tools may produce
+plans which are incorrect or harmful. Applying generated plans without manual
+review is dangerous and may result in the loss of data or service availability.
+Manual review of the plans is required to ensure that infrastructure changes
+are executed safely.
+
+For example, in Terraform one runs `terraform plan -out=path/to/potential.plan`,
+which prints a human-readable plan to `stdout` and produces the file
+`path/to/potential.plan`. A correct workflow requires pausing, reviewing the
+human-readable plan, and only after approving the plan, proceeding to run
+`terraform apply path/to/potential.plan` in order to actually execute the plan
+and make infrastructural changes.
+
+
+# Proposal
+
+This proposal advocates for the addition of a `manual` step to the existing
+variety of steps which may be used in a job (e.g. `get`, `task`, `put`). When
+a `manual` step is added to a job, the Concourse API should expose two options:
+an "approve" option, and a "reject" option. If a user should select the
+"approve" option, then the manual step should be marked as successful and the
+build should proceed normally. If a user should select the "reject" option,
+then the manual step should be marked as failed, and the build should respond
+accordingly, as if any other step in the build had failed.
+
+
+## Supervisors
+
+It is occasionally desirable for the authority to approve `manual` steps to be
+restricted to a limited subset of users. An RBAC role, `supervisor`, should be
+added to Concourse teams. A user with the `supervisor` role has the power to
+approve manual steps. Team admins, by virtue of having the power to alter
```

Comment:
I think the biggest challenge is that team admins could be removing the manual step, invalidating it's usefulness. Any thoughts on that?

Reply:
I don't think that it's reasonable to consider team admins a threat under a security model which depends upon manual steps to enforce policy. A malicious team admin, instead of altering the pipeline to remove the manual step, could do damage by altering the pipeline to add a malicious resource, and more damage would be done, with a lower probability of detection, by doing so. In my opinion, organizations which consider team admins a threat should have separate pipelines, ideally on separate Concourse instances (soon, the same Concourse instance but on different teams, with the introduction of team credential management), such that a different team has the opportunity to review the output from the pipeline which the malicious team admin is administrating.

I see the purpose of a manual step less as a measure to protect against malicious internal actors and more as a measure to review automatically generated output for which automated testing is difficult, or not worth the engineering cost, to implement.